source/2_platform/rex_windows/src/console_application.cpp

```diff
@@ -0,0 +1,114 @@
+#include "rex_windows/console_application.h"
+
+#include "rex_engine/diagnostics/logging/log_macros.h"
+#include "rex_engine/event_system.h"
+
+#include <Windows.h>
+
+namespace rex
+{
+  namespace win32
+  {
+    DEFINE_LOG_CATEGORY(LogWinConsoleApp, rex::LogVerbosity::Log);
+
+    CoreApplication* g_this_app = nullptr; // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+
+    BOOL WINAPI handler_routine(DWORD eventCode)
+    {
+      switch(eventCode) // NOLINT(hicpp-multiway-paths-covered)
+      {
+          // you only have 5 seconds to process this
+          // Windows forcefully kills the app after 5 seconds
+          // see https://learn.microsoft.com/en-us/windows/console/handlerroutine#timeouts
+        case CTRL_CLOSE_EVENT:
+          g_this_app->quit();
+          return false; // NOLINT(readability-implicit-bool-conversion)
+          break;
+      }
+
+      return true; // NOLINT(readability-implicit-bool-conversion)
+    }
+
+    class ConsoleApplication::Internal
+    {
+    public:
+      Internal(CoreApplication* appInstance, ApplicationCreationParams&& appCreationParams)
+          : m_platform_creation_params(rsl::move(appCreationParams.platform_params))
+          , m_cmd_line_args(rsl::move(appCreationParams.cmd_args))
+          , m_engine_params(rsl::move(appCreationParams.engine_params))
+          , m_app_instance(appInstance)
+      {
+        g_this_app = m_app_instance;
+
+        // we're always assigning something to the pointers here to avoid branch checking every update
+        // I've profiled this and always having a function wins here.
+        m_on_initialize = m_engine_params.app_init_func ? m_engine_params.app_init_func : [&]() { return true; };
+
+        m_on_update = m_engine_params.app_update_func ? m_engine_params.app_update_func : [&]() {};
+
+        m_on_shutdown = m_engine_params.app_shutdown_func ? m_engine_params.app_shutdown_func : [&]() {};
+      }
+
+      bool initialize()
+      {
+        SetConsoleCtrlHandler(handler_routine, true); // NOLINT(readability-implicit-bool-conversion)
```

Comment:
Would it be possible to create a constexpr bool here and pass in that to the SetConsoleCtrlHandler, if you give the variable a good name we know what the "true" stands for.